storage/src/main/java/org/apache/kafka/storage/internals/log/LogSegment.java

```diff
@@ -253,17 +253,33 @@ public void append(long largestOffset,
             // append the messages
             long appendedBytes = log.append(records);
             LOGGER.trace("Appended {} to {} at end offset {}", appendedBytes, log.file(), largestOffset);
-            // Update the in memory max timestamp and corresponding offset.
-            if (largestTimestampMs > maxTimestampSoFar()) {
-                maxTimestampAndOffsetSoFar = new TimestampOffset(largestTimestampMs, shallowOffsetOfMaxTimestamp);
-            }
-            // append an entry to the index (if needed)
-            if (bytesSinceLastIndexEntry > indexIntervalBytes) {
-                offsetIndex().append(largestOffset, physicalPosition);
-                timeIndex().maybeAppend(maxTimestampSoFar(), shallowOffsetOfMaxTimestampSoFar());
-                bytesSinceLastIndexEntry = 0;
+
+            long batchMaxTimestamp = RecordBatch.NO_TIMESTAMP;
+            long batchShallowOffsetOfMaxTimestamp = -1L;
+            // append an entry to the index at batches level (if needed)
+            for (RecordBatch batch : records.batches()) {
+                if (batch.maxTimestamp() > batchMaxTimestamp) {
```

Comment:
Could we get rid of this condition and make batchMaxTimestamp and batchShallowOffsetOfMaxTimestamp a local val in the loop?

Reply:
Yes, updated it.